lua/lib/lattice.lua

```diff
@@ -28,10 +28,26 @@ end
 
 --- start running the lattice
 function Lattice:start()
+  self.enabled = true
   if self.auto and self.superclock_id == nil then
     self.superclock_id = clock.run(self.auto_pulse, self)
   end
-  self.enabled = true
+end
+
+--- reset the norns clock and restart lattice
+function Lattice:hard_restart()
+  -- destroy clock, but not the patterns
+  self:stop()
+  if self.superclock_id ~= nil then 
+    clock.cancel(self.superclock_id)
+    self.superclock_id = nil 
+  end
+  for i, pattern in pairs(self.patterns) do
+    self.patterns[i].phase = self.patterns[i].phase_end
+  end
+  self.transport = 0
+  params:set("clock_reset",1)
```

Comment:
what is this doing?

Reply:
`self.patterns[i].phase = self.patterns[i].phase_end` resets patterns by changing their phase to the last phase. that way, when the first new clock pulse hits it will add +1 to the phase and then it will be over the phase limit and trigger the pattern action.

`self.transport = 0` resets the transport so that all the patterns essentially start over with new transports.

`params:set("clock_reset",1)` does the clock reset - basically turning the clock back to 0 and immediately starting it again.

this is all a very much "hard" reset. i could imagine a softer reset going through and not messing with pattern phases, or the clock transports, etc. but I haven't yet found a good use-case for a softer reset (my applications are specifically needing the hard reset actually as I want all the patterns to start over, at the same time).